jena-arq/src/main/java/org/apache/jena/sparql/core/DatasetImpl.java

```diff
@@ -213,11 +248,30 @@ public boolean containsNamedModel(String uri) {
         return dsg.containsGraph(n) ;
     }
 
+    @Override
+    public boolean containsNamedModel(Resource uri) {
+        // Does not touch the cache.
+        checkGraphName(uri) ;
+        return dsg.containsGraph(uri.asNode()) ;
+    }
+
     @Override
     public Iterator<String> listNames() {
```

Comment:
NodeUtils.nodesToURIs does not meet its contract to skip non-URIs.

Here is an impl that does.
```
    public static Iterator<String> nodesToURIs(Iterator<Node> iter) {
        ExtendedIterator<Node> eIter = WrappedIterator.create(iter) ;
        Iterator<String> conv = eIter.filterKeep(Node::isURI).mapWith(Node::getURI);
        return conv ;
    }
```